From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Racci <racci@racci.dev>
Date: Tue, 22 Feb 2022 19:06:45 +1100
Subject: [PATCH] Fire Immunity API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index ffd3d7f4c0c86806bcf70bd90edf3041834d401c..eba00af381979bb7dba30038e608d8832d617922 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -935,4 +935,20 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     boolean isInDaylight();
     // Purpur end
+
+    // Tentacles start
+
+    /**
+     * Checks if the entity is fire immune
+     *
+     * @return True if fire immune
+     */
+    boolean isImmuneToFire();
+
+    /**
+     * Sets if the entity is fire immune
+     * Set this to null to restore the entity type default
+     */
+    void setImmuneToFire(@Nullable Boolean fireImmune);
+    // Tentacles end
 }
diff --git a/src/main/java/org/bukkit/entity/Item.java b/src/main/java/org/bukkit/entity/Item.java
index fd213c026fdfb3030d0c37f651a88857ca038cba..81744c53b2defbbc4cb915968c17b3923220c3a2 100644
--- a/src/main/java/org/bukkit/entity/Item.java
+++ b/src/main/java/org/bukkit/entity/Item.java
@@ -187,7 +187,9 @@ public interface Item extends Entity {
      * Set whether or not this item is immune to fire
      *
      * @param immuneToFire True to make immune to fire
+     * @deprecated use {@link #setImmuneToFire(Boolean)}
      */
+    @Deprecated
     void setImmuneToFire(boolean immuneToFire);
 
     /**
