From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Racci <racci@racci.dev>
Date: Sun, 27 Feb 2022 00:01:09 +1100
Subject: [PATCH] Water Breathing API


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 1da4521fe1930ce66e4d9931d059282d3a66e20d..0525d86463e8f25ce743581cff211b8e2155c598 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -1096,4 +1096,22 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      */
     void setShouldBurnInDay(boolean shouldBurnInDay);
     // Purpur end
+
+    // Tentacles start
+
+    /**
+     * Set if this mob is able to breath underwater
+     * Setting this to null will reset to the mob type default
+     *
+     * @param canBreathUnderwater True to breath underwater
+     */
+    void setCanBreathUnderwater(@Nullable Boolean canBreathUnderwater);
+
+    /**
+     * Get if this mob is able to breath underwater
+     *
+     * @return True if the mob can breath underwater
+     */
+    boolean canBreathUnderwater();
+    // Tentacles end
 }
