From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Racci <racci@racci.dev>
Date: Sun, 27 Feb 2022 00:01:09 +1100
Subject: [PATCH] Water Breathing API


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 6c46107e13a5fd648d64844ffa5b326d4909d702..3275f11e054127417ad8187a8305ee5eea3619d3 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -946,4 +946,22 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      */
     void setShouldBurnInDay(boolean shouldBurnInDay);
     // Purpur end
+
+    // Tentacles start
+
+    /**
+     * Set if this mob is able to breath underwater
+     * Setting this to null will reset to the mob type default
+     *
+     * @param canBreathUnderwater True to breath underwater
+     */
+    void setCanBreathUnderwater(@Nullable Boolean canBreathUnderwater);
+
+    /**
+     * Get if this mob is able to breath underwater
+     *
+     * @return True if the mob can breath underwater
+     */
+    boolean canBreathUnderwater();
+    // Tentacles end
 }
