From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DaRacci <racci@racci.dev>
Date: Thu, 2 Jun 2022 00:44:42 +1000
Subject: [PATCH] Stand on fluid API


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 0525d86463e8f25ce743581cff211b8e2155c598..33ea071f9c0721e32da6a8b3a6510fff7bdbb86e 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -1113,5 +1113,9 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      * @return True if the mob can breath underwater
      */
     boolean canBreathUnderwater();
+
+    void addCanStandOnFluid(@NotNull org.bukkit.Fluid fluid);
+    void removeCanStandOnFluid(@NotNull org.bukkit.Fluid fluid);
+    boolean canStandOnFluid(@NotNull org.bukkit.Fluid fluid);
     // Tentacles end
 }
