From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Racci <racci@racci.dev>
Date: Sun, 27 Feb 2022 00:01:09 +1100
Subject: [PATCH] Water Breathing API


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index bbae010cedf30d0e3336a8f03ee00a0a824cb08d..9604aebbf179e1be85ca648063e8d67d81278d55 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -947,4 +947,22 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      */
     void setShouldBurnInDay(boolean shouldBurnInDay);
     // Purpur end
+
+    // Tentacles start
+
+    /**
+     * Set if this mob is able to breath underwater
+     * Setting this to null will reset to the mob type default
+     *
+     * @param canBreathUnderwater True to breath underwater
+     */
+    void setCanBreathUnderwater(Boolean canBreathUnderwater);
+
+    /**
+     * Get if this mob is able to breath underwater
+     *
+     * @return True if the mob can breath underwater
+     */
+    boolean canBreathUnderwater();
+    // Tentacles end
 }
