From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DaRacci <racci@racci.dev>
Date: Fri, 3 Jun 2022 17:53:04 +1000
Subject: [PATCH] Oxygen-API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 3959765b110b6fed91348c11ff85e302a885776c..55d19ce41d73f0af3c51cfe0a64a5846a3fb94ab 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -1002,4 +1002,17 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     void setImmuneToFire(@Nullable Boolean fireImmune);
     // Purpur end
+
+    // Tentacles start
+    /**
+     * Checks if the entity has reverse oxygen
+     * @return True if gains oxygen in water and looses oxygen in air
+     */
+    boolean isReverseOxygen();
+
+    /**
+     * Sets if the entity has reverse oxygen
+     */
+    void setReverseOxygen(boolean reverseOxygen);
+    // Tentacles end
 }
