From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DaRacci <racci@racci.dev>
Date: Thu, 2 Jun 2022 00:44:42 +1000
Subject: [PATCH] Stand on fluid API


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 9c8466f432811727922101283326321da37f5015..249df330d95d15c9a7712488fa5e33635c82bf2a 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -1040,5 +1040,9 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      * @return True if the mob can breath underwater
      */
     boolean canBreathUnderwater();
+
+    void addCanStandOnFluid(@NotNull org.bukkit.Fluid fluid);
+    void removeCanStandOnFluid(@NotNull org.bukkit.Fluid fluid);
+    boolean canStandOnFluid(@NotNull org.bukkit.Fluid fluid);
     // Tentacles end
 }
