From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DaRacci <racci@racci.dev>
Date: Thu, 2 Jun 2022 00:44:42 +1000
Subject: [PATCH] Stand on fluid API


diff --git a/src/main/java/org/bukkit/entity/LivingEntity.java b/src/main/java/org/bukkit/entity/LivingEntity.java
index 3275f11e054127417ad8187a8305ee5eea3619d3..f5ef33c4e52657cc0324f60c31c99125583306c9 100644
--- a/src/main/java/org/bukkit/entity/LivingEntity.java
+++ b/src/main/java/org/bukkit/entity/LivingEntity.java
@@ -963,5 +963,9 @@ public interface LivingEntity extends Attributable, Damageable, ProjectileSource
      * @return True if the mob can breath underwater
      */
     boolean canBreathUnderwater();
+
+    void addCanStandOnFluid(@NotNull org.bukkit.Fluid fluid);
+    void removeCanStandOnFluid(@NotNull org.bukkit.Fluid fluid);
+    boolean canStandOnFluid(@NotNull org.bukkit.Fluid fluid);
     // Tentacles end
 }
