From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DaRacci <racci@racci.dev>
Date: Fri, 3 Jun 2022 17:53:04 +1000
Subject: [PATCH] Oxygen-API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index c197bd8f291a4a2a5dae5633613aeba13b888ff3..d531415d6f731af8e3820cd4c5cb1c27bb651c0a 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -972,5 +972,16 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      * Set this to null to restore the entity type default
      */
     void setImmuneToFire(@Nullable Boolean fireImmune);
+
+    /**
+     * Checks if the entity has reverse oxygen
+     * @return True if gains oxygen in water and looses oxygen in air
+     */
+    boolean isReverseOxygen();
+
+    /**
+     * Sets if the entity has reverse oxygen
+     */
+    void setReverseOxygen(boolean reverseOxygen);
     // Tentacles end
 }
