From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Racci <racci@racci.dev>
Date: Sun, 27 Feb 2022 00:12:22 +1100
Subject: [PATCH] Water Breathing API


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index f23829690d2be20af50256e4d985e3e342415d54..da01da05aca0f5cda9543d7ab1ef81e9d007ba57 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -266,6 +266,10 @@ public abstract class LivingEntity extends Entity {
     public boolean silentDeath = false; // Paper - mark entity as dying silently for cancellable death event
     public net.kyori.adventure.util.TriState frictionState = net.kyori.adventure.util.TriState.NOT_SET; // Paper
     protected boolean shouldBurnInDay = false;  public boolean shouldBurnInDay() { return this.shouldBurnInDay; } public void setShouldBurnInDay(boolean shouldBurnInDay) { this.shouldBurnInDay = shouldBurnInDay; } // Purpur
+    // Tentacles start
+    @Nullable
+    public Boolean canBreathUnderwater = null;
+    // Tentacles end
 
     @Override
     public float getBukkitYaw() {
@@ -378,7 +382,7 @@ public abstract class LivingEntity extends Entity {
     }
 
     public boolean canBreatheUnderwater() {
-        return this.getMobType() == MobType.UNDEAD;
+        return canBreathUnderwater != null ? canBreathUnderwater : this.getMobType() == MobType.UNDEAD;
     }
 
     public float getSwimAmount(float tickDelta) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 7d4c8d205fa7fdc64358ad8fac6fa7d7fc9166a0..d5faed7430522884560e301fabb59b30ccc934a7 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -1071,4 +1071,16 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         getHandle().setShouldBurnInDay(shouldBurnInDay);
     }
     // Purpur end
+
+    // Tentacles start
+    @Override
+    public void setCanBreathUnderwater(Boolean canBreathUnderwater) {
+        getHandle().canBreathUnderwater = canBreathUnderwater;
+    }
+
+    @Override
+    public boolean canBreathUnderwater() {
+        return getHandle().canBreatheUnderwater();
+    }
+    // Tentacles end
 }
