From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Racci <racci@racci.dev>
Date: Sun, 27 Feb 2022 00:12:22 +1100
Subject: [PATCH] Water Breathing API


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 1ec9a48f2f44f5bd647e0661423ae90970aee262..d5fcafbcdc1264b4b491936cf71ce634a11a2450 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -265,6 +265,10 @@ public abstract class LivingEntity extends Entity {
     public org.bukkit.craftbukkit.entity.CraftLivingEntity getBukkitLivingEntity() { return (org.bukkit.craftbukkit.entity.CraftLivingEntity) super.getBukkitEntity(); } // Paper
     public boolean silentDeath = false; // Paper - mark entity as dying silently for cancellable death event
     protected boolean shouldBurnInDay = false;  public boolean shouldBurnInDay() { return this.shouldBurnInDay; } public void setShouldBurnInDay(boolean shouldBurnInDay) { this.shouldBurnInDay = shouldBurnInDay; } // Purpur
+    // Tentacles start
+    @Nullable
+    public Boolean canBreathUnderwater = null;
+    // Tentacles end
 
     @Override
     public float getBukkitYaw() {
@@ -377,7 +381,7 @@ public abstract class LivingEntity extends Entity {
     }
 
     public boolean canBreatheUnderwater() {
-        return this.getMobType() == MobType.UNDEAD;
+        return canBreathUnderwater != null ? canBreathUnderwater : this.getMobType() == MobType.UNDEAD;
     }
 
     public float getSwimAmount(float tickDelta) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index cdfdcaa296a685f23b7aea683d68cb794ac1e0c7..6aa4ff45a30e32464bde16318a35340ff8c2b91f 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -1012,4 +1012,16 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         getHandle().setShouldBurnInDay(shouldBurnInDay);
     }
     // Purpur end
+
+    // Tentacles start
+    @Override
+    public void setCanBreathUnderwater(Boolean canBreathUnderwater) {
+        getHandle().canBreathUnderwater = canBreathUnderwater;
+    }
+
+    @Override
+    public boolean canBreathUnderwater() {
+        return getHandle().canBreatheUnderwater();
+    }
+    // Tentacles end
 }
