From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Racci <racci@racci.dev>
Date: Sun, 27 Feb 2022 00:12:22 +1100
Subject: [PATCH] Water Breathing API


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 411593b1b105d62440d76b7bd1b8c74b701e3e75..96fa8268527e6095ad4f8229368fa9b5489073d5 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -266,6 +266,10 @@ public abstract class LivingEntity extends Entity {
     public org.bukkit.craftbukkit.entity.CraftLivingEntity getBukkitLivingEntity() { return (org.bukkit.craftbukkit.entity.CraftLivingEntity) super.getBukkitEntity(); } // Paper
     public boolean silentDeath = false; // Paper - mark entity as dying silently for cancellable death event
     protected boolean shouldBurnInDay = false;  public boolean shouldBurnInDay() { return this.shouldBurnInDay; } public void setShouldBurnInDay(boolean shouldBurnInDay) { this.shouldBurnInDay = shouldBurnInDay; } // Purpur
+    // Tentacles start
+    @Nullable
+    public Boolean canBreathUnderwater = null;
+    // Tentacles end
 
     @Override
     public float getBukkitYaw() {
@@ -378,7 +382,7 @@ public abstract class LivingEntity extends Entity {
     }
 
     public boolean canBreatheUnderwater() {
-        return this.getMobType() == MobType.UNDEAD;
+        return canBreathUnderwater != null ? canBreathUnderwater : this.getMobType() == MobType.UNDEAD;
     }
 
     public float getSwimAmount(float tickDelta) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 749bcdd5ac8dbe9333e74068ff577eabf3a8da29..96337ab39257aef77491db310d92f3a7e6168c0e 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -948,4 +948,16 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         getHandle().setShouldBurnInDay(shouldBurnInDay);
     }
     // Purpur end
+
+    // Tentacles start
+    @Override
+    public void setCanBreathUnderwater(Boolean canBreathUnderwater) {
+        getHandle().canBreathUnderwater = canBreathUnderwater;
+    }
+
+    @Override
+    public boolean canBreathUnderwater() {
+        return getHandle().canBreatheUnderwater();
+    }
+    // Tentacles end
 }
