From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Racci <racci@racci.dev>
Date: Sun, 27 Feb 2022 00:12:22 +1100
Subject: [PATCH] Water Breathing API


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 8f80b2f14c582d248988a824193d1fd0df177dc5..ac33ad2de88bc5d5f0335c488c05975761ff8cb0 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -266,6 +266,10 @@ public abstract class LivingEntity extends Entity {
     public org.bukkit.craftbukkit.entity.CraftLivingEntity getBukkitLivingEntity() { return (org.bukkit.craftbukkit.entity.CraftLivingEntity) super.getBukkitEntity(); } // Paper
     public boolean silentDeath = false; // Paper - mark entity as dying silently for cancellable death event
     protected boolean shouldBurnInDay = false;  public boolean shouldBurnInDay() { return this.shouldBurnInDay; } public void setShouldBurnInDay(boolean shouldBurnInDay) { this.shouldBurnInDay = shouldBurnInDay; } // Purpur
+    // Tentacles start
+    @Nullable
+    public Boolean canBreathUnderwater = null;
+    // Tentacles end
 
     @Override
     public float getBukkitYaw() {
@@ -378,7 +382,7 @@ public abstract class LivingEntity extends Entity {
     }
 
     public boolean canBreatheUnderwater() {
-        return this.getMobType() == MobType.UNDEAD;
+        return canBreathUnderwater != null ? canBreathUnderwater : this.getMobType() == MobType.UNDEAD;
     }
 
     public float getSwimAmount(float tickDelta) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 81ec8f10ed49cef6a620483d1dd828f7b509a8e5..59357e21840a3890fca00fca77ed24ceaf7777c3 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -994,4 +994,16 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         getHandle().setShouldBurnInDay(shouldBurnInDay);
     }
     // Purpur end
+
+    // Tentacles start
+    @Override
+    public void setCanBreathUnderwater(Boolean canBreathUnderwater) {
+        getHandle().canBreathUnderwater = canBreathUnderwater;
+    }
+
+    @Override
+    public boolean canBreathUnderwater() {
+        return getHandle().canBreatheUnderwater();
+    }
+    // Tentacles end
 }
