From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DaRacci <racci@racci.dev>
Date: Fri, 3 Jun 2022 17:53:09 +1000
Subject: [PATCH] Oxygen-API


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 967647c7dc0a4c2d762e7133fe969106f7d8ecba..2a2f7252d5e1967ae0c2fcc42dfdb1b7abd2f63c 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -340,7 +340,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     private static final int FLAG_INVISIBLE = 5;
     protected static final int FLAG_GLOWING = 6;
     protected static final int FLAG_FALL_FLYING = 7;
-    private static final EntityDataAccessor<Integer> DATA_AIR_SUPPLY_ID = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.INT);
+    protected static final EntityDataAccessor<Integer> DATA_AIR_SUPPLY_ID = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.INT); // Tentacles - private->protected
     private static final EntityDataAccessor<Optional<Component>> DATA_CUSTOM_NAME = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.OPTIONAL_COMPONENT);
     private static final EntityDataAccessor<Boolean> DATA_CUSTOM_NAME_VISIBLE = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.BOOLEAN);
     private static final EntityDataAccessor<Boolean> DATA_SILENT = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.BOOLEAN);
@@ -399,6 +399,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     private UUID originWorld;
     public boolean freezeLocked = false; // Paper - Freeze Tick Lock API
     public @Nullable Boolean immuneToFire = null; // Purpur - Fire immune API
+    public boolean reverseOxygen = false; // Tentacles - Oxygen API
 
     public void setOrigin(@javax.annotation.Nonnull Location location) {
         this.origin = location.toVector();
@@ -2307,6 +2308,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
                 nbt.putBoolean("Purpur.FireImmune", immuneToFire);
             }
             // Purpur end
+            // Tentacles start
+            if (reverseOxygen) {
+                nbt.putBoolean("Tentacles.ReverseOxygen", true);
+            }
+            // Tentacles end
             return nbt;
         } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.forThrowable(throwable, "Saving entity NBT");
@@ -2479,6 +2485,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
                 immuneToFire = nbt.getBoolean("Purpur.FireImmune");
             }
             // Purpur end
+            // Tentacles start
+            if (nbt.contains("Tentacles.ReverseOxygen")) {
+                reverseOxygen = nbt.getBoolean("Tentacles.ReverseOxygen");
+            }
+            // Tentacles end
 
         } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.forThrowable(throwable, "Loading entity NBT");
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 5b25c686aab37af1ea835116b1d28e3ebc1bb745..c41e4af52658012632f08f291eaa5556aa733d36 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -427,28 +427,24 @@ public abstract class LivingEntity extends Entity {
                 boolean flag1 = !this.canBreatheUnderwater() && !MobEffectUtil.hasWaterBreathing(this) && (!flag || !((net.minecraft.world.entity.player.Player) this).getAbilities().invulnerable);
 
                 if (flag1) {
-                    this.setAirSupply(this.decreaseAirSupply(this.getAirSupply()));
-                    if (this.getAirSupply() == -this.level.purpurConfig.drowningDamageInterval) { // Purpur
-                        this.setAirSupply(0);
-                        Vec3 vec3d = this.getDeltaMovement();
-
-                        for (int i = 0; i < 8; ++i) {
-                            double d2 = this.random.nextDouble() - this.random.nextDouble();
-                            double d3 = this.random.nextDouble() - this.random.nextDouble();
-                            double d4 = this.random.nextDouble() - this.random.nextDouble();
-
-                            this.level.addParticle(ParticleTypes.BUBBLE, this.getX() + d2, this.getY() + d3, this.getZ() + d4, vec3d.x, vec3d.y, vec3d.z);
-                        }
-
-                        this.hurt(DamageSource.DROWN, (float) this.level.purpurConfig.damageFromDrowning); // Purpur
+                    // Tentacles start
+                    if (reverseOxygen) {
+                        this.increaseAir();
+                    } else {
+                        this.decreaseAir();
                     }
+                    // Tentacles end
                 }
 
                 if (!this.level.isClientSide && this.isPassenger() && this.getVehicle() != null && !this.getVehicle().rideableUnderWater()) {
                     this.stopRiding();
                 }
-            } else if (this.getAirSupply() < this.getMaxAirSupply()) {
-                this.setAirSupply(this.increaseAirSupply(this.getAirSupply()));
+                // Tentacles start
+            } else if (reverseOxygen) {
+                this.decreaseAir();
+            } else {
+                this.increaseAir();
+                // Tentacles end
             }
 
             if (!this.level.isClientSide) {
@@ -508,6 +504,32 @@ public abstract class LivingEntity extends Entity {
         //this.level.getProfiler().pop(); // Purpur
     }
 
+    // Tentacles start - Extract method
+    private void increaseAir() {
+        if (this.getAirSupply() < this.getMaxAirSupply()) {
+            this.setAirSupply(this.increaseAirSupply(this.getAirSupply()));
+        }
+    }
+
+    private void decreaseAir() {
+        this.setAirSupply(this.decreaseAirSupply(this.getAirSupply()));
+        if (this.getAirSupply() == -this.level.purpurConfig.drowningDamageInterval) { // Purpur
+            this.setAirSupply(0);
+            Vec3 vec3d = this.getDeltaMovement();
+
+            for (int i = 0; i < 8; ++i) {
+                double d2 = this.random.nextDouble() - this.random.nextDouble();
+                double d3 = this.random.nextDouble() - this.random.nextDouble();
+                double d4 = this.random.nextDouble() - this.random.nextDouble();
+
+                this.level.addParticle(ParticleTypes.BUBBLE, this.getX() + d2, this.getY() + d3, this.getZ() + d4, vec3d.x, vec3d.y, vec3d.z);
+            }
+
+            this.hurt(DamageSource.DROWN, (float) this.level.purpurConfig.damageFromDrowning); // Purpur
+        }
+    }
+    // Tentacles end
+
     public boolean canSpawnSoulSpeedParticle() {
         return this.tickCount % 5 == 0 && this.getDeltaMovement().x != 0.0D && this.getDeltaMovement().z != 0.0D && !this.isSpectator() && EnchantmentHelper.hasSoulSpeed(this) && this.onSoulSpeedBlock();
     }
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index e87a436711e179a4384630de41188c09a75a7f0d..49831829bb7f75ce659d61af167cd1f034fc6212 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -305,6 +305,11 @@ public abstract class Player extends LivingEntity {
         }
 
         this.updateIsUnderwater();
+        // Tentacles start - Client tries to reduce air on its own
+        if (this.reverseOxygen && this.isUnderWater() && this.getAirSupply() == this.getMaxAirSupply()) {
+            this.entityData.markDirty(Entity.DATA_AIR_SUPPLY_ID);
+        }
+        // Tentacles end
         super.tick();
         if (!this.level.isClientSide && this.containerMenu != null && !this.containerMenu.stillValid(this)) {
             this.closeContainer(org.bukkit.event.inventory.InventoryCloseEvent.Reason.CANT_USE); // Paper
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 10208b3be330e8a02a4b248a3861ba978e64aab4..8875d75960ad35858bfea491e87c7d96d737f404 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1426,4 +1426,16 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         getHandle().immuneToFire = fireImmune;
     }
     // Purpur end
+
+    // Tentacles start
+    @Override
+    public boolean isReverseOxygen() {
+        return getHandle().reverseOxygen;
+    }
+
+    @Override
+    public void setReverseOxygen(boolean reverseOxygen) {
+        getHandle().reverseOxygen = reverseOxygen;
+    }
+    // Tentacles end
 }
