From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DaRacci <racci@racci.dev>
Date: Fri, 3 Jun 2022 17:53:09 +1000
Subject: [PATCH] Oxygen-API


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 4dbedb5192983dfa42c3de6ec7b754ed8a1253be..2791bb18ec509d643602d3ffcf3ca9ffcd5cdcce 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -341,7 +341,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     private static final int FLAG_INVISIBLE = 5;
     protected static final int FLAG_GLOWING = 6;
     protected static final int FLAG_FALL_FLYING = 7;
-    private static final EntityDataAccessor<Integer> DATA_AIR_SUPPLY_ID = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.INT);
+    protected static final EntityDataAccessor<Integer> DATA_AIR_SUPPLY_ID = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.INT); // Tentacles - private->protected
     private static final EntityDataAccessor<Optional<Component>> DATA_CUSTOM_NAME = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.OPTIONAL_COMPONENT);
     private static final EntityDataAccessor<Boolean> DATA_CUSTOM_NAME_VISIBLE = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.BOOLEAN);
     private static final EntityDataAccessor<Boolean> DATA_SILENT = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.BOOLEAN);
@@ -400,6 +400,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     private UUID originWorld;
     public boolean freezeLocked = false; // Paper - Freeze Tick Lock API
     public @Nullable Boolean immuneToFire = null; // Tentacles - Fire Immunity API
+    public boolean reverseOxygen = true; // Tentacles - Oxygen API
 
     public void setOrigin(@javax.annotation.Nonnull Location location) {
         this.origin = location.toVector();
@@ -2292,6 +2293,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
             if (immuneToFire != null) {
                 nbt.putBoolean("Purpur.FireImmune", immuneToFire);
             }
+            if (reverseOxygen) {
+                nbt.putBoolean("Tentacles.ReverseOxygen", true);
+            }
             // Tentacles end
             return nbt;
         } catch (Throwable throwable) {
@@ -2464,6 +2468,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
             if (nbt.contains("Purpur.FireImmune")) {
                 immuneToFire = nbt.getBoolean("Purpur.FireImmune");
             }
+            if (nbt.contains("Tentacles.ReverseOxygen")) {
+                reverseOxygen = nbt.getBoolean("Tentacles.ReverseOxygen");
+            }
             // Tentacles end
 
         } catch (Throwable throwable) {
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 97af52d449ff2e7b0ca7744db5efb32979abce9c..757d610913a5e22173f36c5214d4a6f3108f3d3f 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -428,28 +428,24 @@ public abstract class LivingEntity extends Entity {
                 boolean flag1 = !this.canBreatheUnderwater() && !MobEffectUtil.hasWaterBreathing(this) && (!flag || !((net.minecraft.world.entity.player.Player) this).getAbilities().invulnerable);
 
                 if (flag1) {
-                    this.setAirSupply(this.decreaseAirSupply(this.getAirSupply()));
-                    if (this.getAirSupply() == -this.level.purpurConfig.drowningDamageInterval) { // Purpur
-                        this.setAirSupply(0);
-                        Vec3 vec3d = this.getDeltaMovement();
-
-                        for (int i = 0; i < 8; ++i) {
-                            double d2 = this.random.nextDouble() - this.random.nextDouble();
-                            double d3 = this.random.nextDouble() - this.random.nextDouble();
-                            double d4 = this.random.nextDouble() - this.random.nextDouble();
-
-                            this.level.addParticle(ParticleTypes.BUBBLE, this.getX() + d2, this.getY() + d3, this.getZ() + d4, vec3d.x, vec3d.y, vec3d.z);
-                        }
-
-                        this.hurt(DamageSource.DROWN, (float) this.level.purpurConfig.damageFromDrowning); // Purpur
+                    // Tentacles start
+                    if (reverseOxygen) {
+                        this.increaseAir();
+                    } else {
+                        this.decreaseAir();
                     }
+                    // Tentacles end
                 }
 
                 if (!this.level.isClientSide && this.isPassenger() && this.getVehicle() != null && !this.getVehicle().rideableUnderWater()) {
                     this.stopRiding();
                 }
-            } else if (this.getAirSupply() < this.getMaxAirSupply()) {
-                this.setAirSupply(this.increaseAirSupply(this.getAirSupply()));
+                // Tentacles start
+            } else if (reverseOxygen) {
+                this.decreaseAir();
+            } else {
+                this.increaseAir();
+                // Tentacles end
             }
 
             if (!this.level.isClientSide) {
@@ -509,6 +505,32 @@ public abstract class LivingEntity extends Entity {
         //this.level.getProfiler().pop(); // Purpur
     }
 
+    // Tentacles start - Extract method
+    private void increaseAir() {
+        if (this.getAirSupply() < this.getMaxAirSupply()) {
+            this.setAirSupply(this.increaseAirSupply(this.getAirSupply()));
+        }
+    }
+
+    private void decreaseAir() {
+        this.setAirSupply(this.decreaseAirSupply(this.getAirSupply()));
+        if (this.getAirSupply() == -this.level.purpurConfig.drowningDamageInterval) { // Purpur
+            this.setAirSupply(0);
+            Vec3 vec3d = this.getDeltaMovement();
+
+            for (int i = 0; i < 8; ++i) {
+                double d2 = this.random.nextDouble() - this.random.nextDouble();
+                double d3 = this.random.nextDouble() - this.random.nextDouble();
+                double d4 = this.random.nextDouble() - this.random.nextDouble();
+
+                this.level.addParticle(ParticleTypes.BUBBLE, this.getX() + d2, this.getY() + d3, this.getZ() + d4, vec3d.x, vec3d.y, vec3d.z);
+            }
+
+            this.hurt(DamageSource.DROWN, (float) this.level.purpurConfig.damageFromDrowning); // Purpur
+        }
+    }
+    // Tentacles end
+
     public boolean canSpawnSoulSpeedParticle() {
         return this.tickCount % 5 == 0 && this.getDeltaMovement().x != 0.0D && this.getDeltaMovement().z != 0.0D && !this.isSpectator() && EnchantmentHelper.hasSoulSpeed(this) && this.onSoulSpeedBlock();
     }
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 64ca2c92c4908bbb0272d1c22870af68aa9daad2..a13d69a11426157f5c3c31227c20f7413f55d44f 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -305,6 +305,11 @@ public abstract class Player extends LivingEntity {
         }
 
         this.updateIsUnderwater();
+        // Tentacles start - Client tries to reduce air on its own
+        if (this.reverseOxygen && this.isUnderWater() && this.getAirSupply() == this.getMaxAirSupply()) {
+            this.entityData.markDirty(Entity.DATA_AIR_SUPPLY_ID);
+        }
+        // Tentacles end
         super.tick();
         if (!this.level.isClientSide && this.containerMenu != null && !this.containerMenu.stillValid(this)) {
             this.closeContainer(org.bukkit.event.inventory.InventoryCloseEvent.Reason.CANT_USE); // Paper
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 2fc7fb7c907b8e05b66bdd834e5191bc10ec6d4b..d28502200aefc10db1e8be5a29459e12413fa969 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1398,5 +1398,15 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     public void setImmuneToFire(Boolean fireImmune) {
         getHandle().immuneToFire = fireImmune;
     }
+
+    @Override
+    public boolean isReverseOxygen() {
+        return getHandle().reverseOxygen;
+    }
+
+    @Override
+    public void setReverseOxygen(boolean reverseOxygen) {
+        getHandle().reverseOxygen = reverseOxygen;
+    }
     // Tentacles end
 }
