From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DaRacci <racci@racci.dev>
Date: Fri, 3 Jun 2022 17:53:09 +1000
Subject: [PATCH] Oxygen-API


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 684c81d692492cc0aa492b12fec26cead8821683..4946dd3dedb040e8811a68030780041570ab094b 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -399,6 +399,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     private UUID originWorld;
     public boolean freezeLocked = false; // Paper - Freeze Tick Lock API
     public @Nullable Boolean immuneToFire = null; // Tentacles - Fire Immunity API
+    public boolean reverseOxygen = false; // Tentacles - Oxygen API
 
     public void setOrigin(@javax.annotation.Nonnull Location location) {
         this.origin = location.toVector();
@@ -2297,6 +2298,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
             if (immuneToFire != null) {
                 nbt.putBoolean("Purpur.FireImmune", immuneToFire);
             }
+            if (reverseOxygen) {
+                nbt.putBoolean("Tentacles.ReverseOxygen", true);
+            }
             // Tentacles end
             return nbt;
         } catch (Throwable throwable) {
@@ -2467,6 +2471,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
             if (nbt.contains("Purpur.FireImmune")) {
                 immuneToFire = nbt.getBoolean("Purpur.FireImmune");
             }
+            if (nbt.contains("Tentacles.ReverseOxygen")) {
+                reverseOxygen = nbt.getBoolean("Tentacles.ReverseOxygen");
+            }
             // Tentacles end
 
         } catch (Throwable throwable) {
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 0434beb0f473bc1292b9e78879ec3a0b5986b000..a5abf59f48d04985b0fe714ff96158d9215c797a 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -428,28 +428,24 @@ public abstract class LivingEntity extends Entity {
                 boolean flag1 = !this.canBreatheUnderwater() && !MobEffectUtil.hasWaterBreathing(this) && (!flag || !((net.minecraft.world.entity.player.Player) this).getAbilities().invulnerable);
 
                 if (flag1) {
-                    this.setAirSupply(this.decreaseAirSupply(this.getAirSupply()));
-                    if (this.getAirSupply() == -this.level.purpurConfig.drowningDamageInterval) { // Purpur
-                        this.setAirSupply(0);
-                        Vec3 vec3d = this.getDeltaMovement();
-
-                        for (int i = 0; i < 8; ++i) {
-                            double d2 = this.random.nextDouble() - this.random.nextDouble();
-                            double d3 = this.random.nextDouble() - this.random.nextDouble();
-                            double d4 = this.random.nextDouble() - this.random.nextDouble();
-
-                            this.level.addParticle(ParticleTypes.BUBBLE, this.getX() + d2, this.getY() + d3, this.getZ() + d4, vec3d.x, vec3d.y, vec3d.z);
-                        }
-
-                        this.hurt(DamageSource.DROWN, (float) this.level.purpurConfig.damageFromDrowning); // Purpur
+                    // Tentacles start
+                    if (reverseOxygen) {
+                        this.increaseAir();
+                    } else {
+                        this.decreaseAir();
                     }
+                    // Tentacles end
                 }
 
                 if (!this.level.isClientSide && this.isPassenger() && this.getVehicle() != null && !this.getVehicle().rideableUnderWater()) {
                     this.stopRiding();
                 }
-            } else if (this.getAirSupply() < this.getMaxAirSupply()) {
-                this.setAirSupply(this.increaseAirSupply(this.getAirSupply()));
+                // Tentacles start
+            } else if (reverseOxygen) {
+                this.decreaseAir();
+            } else {
+                this.increaseAir();
+                // Tentacles end
             }
 
             if (!this.level.isClientSide) {
@@ -509,6 +505,32 @@ public abstract class LivingEntity extends Entity {
         this.level.getProfiler().pop();
     }
 
+    // Tentacles start - Extract method
+    private void increaseAir() {
+        if (this.getAirSupply() < this.getMaxAirSupply()) {
+            this.setAirSupply(this.increaseAirSupply(this.getAirSupply()));
+        }
+    }
+
+    private void decreaseAir() {
+        this.setAirSupply(this.decreaseAirSupply(this.getAirSupply()));
+        if (this.getAirSupply() == -this.level.purpurConfig.drowningDamageInterval) { // Purpur
+            this.setAirSupply(0);
+            Vec3 vec3d = this.getDeltaMovement();
+
+            for (int i = 0; i < 8; ++i) {
+                double d2 = this.random.nextDouble() - this.random.nextDouble();
+                double d3 = this.random.nextDouble() - this.random.nextDouble();
+                double d4 = this.random.nextDouble() - this.random.nextDouble();
+
+                this.level.addParticle(ParticleTypes.BUBBLE, this.getX() + d2, this.getY() + d3, this.getZ() + d4, vec3d.x, vec3d.y, vec3d.z);
+            }
+
+            this.hurt(DamageSource.DROWN, (float) this.level.purpurConfig.damageFromDrowning); // Purpur
+        }
+    }
+    // Tentacles end
+
     public boolean canSpawnSoulSpeedParticle() {
         return this.tickCount % 5 == 0 && this.getDeltaMovement().x != 0.0D && this.getDeltaMovement().z != 0.0D && !this.isSpectator() && EnchantmentHelper.hasSoulSpeed(this) && this.onSoulSpeedBlock();
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 94e9ba12a357ed6010bcd404f3a6c044e7b8e340..8a22ef6065b6cf7d8c2f6f208969836fd864adc3 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1365,5 +1365,15 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     public void setImmuneToFire(Boolean fireImmune) {
         getHandle().immuneToFire = fireImmune;
     }
+
+    @Override
+    public boolean isReverseOxygen() {
+        return getHandle().reverseOxygen;
+    }
+
+    @Override
+    public void setReverseOxygen(boolean reverseOxygen) {
+        getHandle().reverseOxygen = reverseOxygen;
+    }
     // Tentacles end
 }
