From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DaRacci <racci@racci.dev>
Date: Sat, 5 Nov 2022 17:26:03 +1100
Subject: [PATCH] Mining Speed API


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 467f60dcbed645d773e7e3e042d2bbbd8f5bb589..b68ceb3c7f912d3dbf2c213b9295cbef994629fa 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -6,12 +6,19 @@ import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.util.Either;
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.DataResult;
+import dev.racci.tentacles.Tentacles;
+import it.unimi.dsi.fastutil.Pair; // Tentacles
 import java.util.Collection;
+import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Map; // Tentacles
 import java.util.Objects;
 import java.util.Optional;
 import java.util.OptionalInt;
+import java.util.concurrent.atomic.AtomicBoolean; // Tentacles
+import java.util.concurrent.atomic.AtomicReference; // Tentacles
+import java.util.function.BiPredicate;
 import javax.annotation.Nullable;
 import net.minecraft.BlockUtil;
 import net.minecraft.ChatFormatting;
@@ -89,12 +96,14 @@ import net.minecraft.stats.ServerRecipeBook;
 import net.minecraft.stats.ServerStatsCounter;
 import net.minecraft.stats.Stat;
 import net.minecraft.stats.Stats;
+import net.minecraft.tags.FluidTags; // Tentacles
 import net.minecraft.util.Mth;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.Unit;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.damagesource.EntityDamageSource;
 import net.minecraft.world.effect.MobEffectInstance;
+import net.minecraft.world.effect.MobEffectUtil; // Tentacles
 import net.minecraft.world.effect.MobEffects;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntitySelector;
@@ -138,6 +147,7 @@ import net.minecraft.world.level.block.NetherPortalBlock;
 import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.entity.CommandBlockEntity;
 import net.minecraft.world.level.block.entity.SignBlockEntity;
+import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.border.WorldBorder;
 import net.minecraft.world.level.dimension.LevelStem;
@@ -145,6 +155,10 @@ import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.portal.PortalInfo;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.NamespacedKey; // Tentacles
+import org.bukkit.craftbukkit.block.CraftBlockState;
+import org.bukkit.craftbukkit.block.CraftBlockStates;
+import org.bukkit.craftbukkit.block.data.CraftBlockData;
 import org.slf4j.Logger;
 import net.minecraft.world.Container;
 import net.minecraft.world.InteractionHand;
@@ -271,6 +285,7 @@ public class ServerPlayer extends Player {
     public double lastEntitySpawnRadiusSquared; // Paper - optimise isOutsideRange, this field is in blocks
     public final com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<ServerPlayer> cachedSingleHashSet; // Paper
     public org.bukkit.event.player.PlayerQuitEvent.QuitReason quitReason = null; // Paper - there are a lot of changes to do if we change all methods leading to the event
+    public boolean alwaysCorrectTool = false; // Tentacles - Mining speed API
 
     public ServerPlayer(MinecraftServer server, ServerLevel world, GameProfile profile, @Nullable ProfilePublicKey publicKey) {
         super(world, world.getSharedSpawnPos(), world.getSharedSpawnAngle(), profile, publicKey);
@@ -2667,4 +2682,26 @@ public class ServerPlayer extends Player {
         this.compassBar = compassBar;
     }
     // Purpur end
+
+    // Tentacles start
+    @Override
+    public float getDestroySpeed(BlockState state) {
+        return this.getDestroySpeedPair(state).left();
+    }
+
+    public Pair<Float, Boolean> getDestroySpeedPair(BlockState block) {
+        final var baseSpeed = super.getDestroySpeed(block);
+        final var finalDestroySpeed = new AtomicReference<>((float) baseSpeed);
+
+        final var bukkitState = CraftBlockStates.getBlockState(block, null);
+        final var bukkitPlayer = this.getBukkitEntity();
+
+        Tentacles.getGlobalMiningModifiers().values().stream()
+            .map(function -> function.apply(bukkitPlayer, bukkitState))
+            .filter(Objects::nonNull)
+            .forEach(modifier -> finalDestroySpeed.updateAndGet(speed -> speed * modifier));
+
+        return Pair.of(finalDestroySpeed.get(), baseSpeed != finalDestroySpeed.get());
+    }
+    // Tentacles end
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
index 4816b5bd54ae6e080af4c19df43d57e852e76146..dd7f998ac58febe8d71a8ff206b144460ba0bedc 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -3,6 +3,12 @@ package net.minecraft.server.level;
 import com.mojang.logging.LogUtils;
 import java.util.Objects;
 import javax.annotation.Nullable;
+import net.minecraft.core.particles.BlockParticleOption; // Tentacles
+import net.minecraft.core.particles.ParticleTypes; // Tentacles
+import net.minecraft.network.protocol.game.ClientboundBlockDestructionPacket;
+import net.minecraft.network.protocol.game.ClientboundLevelParticlesPacket; // Tentacles
+import net.minecraft.network.protocol.game.ClientboundSoundPacket; // Tentacles
+import net.minecraft.sounds.SoundSource; // Tentacles
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.InteractionResultHolder;
@@ -64,6 +70,7 @@ public class ServerPlayerGameMode {
     private int lastSentState;
     public boolean captureSentBlockEntities = false; // Paper
     public boolean capturedBlockEntity = false; // Paper
+    public boolean manipulatedMining = false; // Tentacles
 
     public ServerPlayerGameMode(ServerPlayer player) {
         this.gameModeForPlayer = GameType.DEFAULT_MODE;
@@ -135,6 +142,17 @@ public class ServerPlayerGameMode {
 
                 if (f >= 1.0F) {
                     this.hasDelayedDestroy = false;
+                    // Tentacles start - Mining speed API
+                    if (manipulatedMining) {
+                        final var x = delayedDestroyPos.getX() + 0.5;
+                        final var y = delayedDestroyPos.getY() + 0.5;
+                        final var z = delayedDestroyPos.getZ() + 0.5;
+
+                        this.player.connection.send(new ClientboundLevelParticlesPacket(new BlockParticleOption(ParticleTypes.BLOCK, iblockdata), false, x, y, z, 0.1F, 0.1F, 0.1F, 4, 100));
+                        this.player.connection.send(new ClientboundSoundPacket(iblockdata.getSoundType().breakSound, SoundSource.BLOCKS, x, y, z, 1.0F, 1.0F, this.level.getSeed()));
+                        this.manipulatedMining = false;
+                    }
+                    // Tentacles end
                     this.destroyBlock(this.delayedDestroyPos);
                 }
             }
@@ -152,6 +170,23 @@ public class ServerPlayerGameMode {
                 this.isDestroyingBlock = false;
             } else {
                 this.incrementDestroyProgress(iblockdata, this.destroyPos, this.destroyProgressStart);
+
+                if (this.manipulatedMining && !this.hasDelayedDestroy) {
+                    final int miningTime = this.gameTicks - this.destroyProgressStart;
+                    final var blockData = this.level.getBlockState(this.destroyPos);
+
+                    if (blockData.isAir()) return;
+
+                    final float miningProgress = blockData.getDestroyProgress(this.player, this.player.level, this.destroyPos) * (miningTime + 1);
+
+                    if (miningProgress >= 1.0F) {
+                        this.manipulatedMining = false;
+                        this.isDestroyingBlock = false;
+                        this.hasDelayedDestroy = true;
+                        this.delayedDestroyPos = this.destroyPos;
+                        this.delayedTickStart = this.destroyProgressStart;
+                    }
+                }
             }
         }
 
@@ -159,14 +194,20 @@ public class ServerPlayerGameMode {
 
     private float incrementDestroyProgress(BlockState state, BlockPos pos, int failedStartMiningTime) {
         int j = this.gameTicks - failedStartMiningTime;
-        float f = state.getDestroyProgress(this.player, this.player.level, pos) * (float) (j + 1);
+        // Tentacles start - Mining speed API
+        final var pair = state.getDestroyProgressPair(this.player, this.player.level, pos);
+        float f = pair.left() * (float) (j + 1);
         int k = (int) (f * 10.0F);
 
         if (k != this.lastSentState) {
             this.level.destroyBlockProgress(this.player.getId(), pos, k);
             this.lastSentState = k;
+            if (pair.right()) this.player.connection.send(new ClientboundBlockDestructionPacket(this.player.getId() - 1, pos, k));
         }
 
+        if (pair.right()) this.manipulatedMining = true;
+        // Tentacles end
+
         return f;
     }
 
@@ -294,6 +335,7 @@ public class ServerPlayerGameMode {
                         }
 
                         if (!this.hasDelayedDestroy) {
+                            this.manipulatedMining = false;
                             this.isDestroyingBlock = false;
                             this.hasDelayedDestroy = true;
                             this.delayedDestroyPos = pos;
@@ -310,9 +352,17 @@ public class ServerPlayerGameMode {
                     BlockState type = this.level.getBlockStateIfLoaded(this.destroyPos); // Paper - don't load unloaded chunks for stale records here
                     if (type != null) this.level.destroyBlockProgress(this.player.getId(), this.destroyPos, -1);
                     if (type != null) this.debugLogging(pos, true, sequence, "aborted mismatched destroying");
+
                     this.destroyPos = BlockPos.ZERO; // Paper
                 }
 
+                // Tentacles start - Mining speed API
+                if (manipulatedMining) {
+                    this.manipulatedMining = false;
+                    this.player.connection.send(new ClientboundBlockDestructionPacket(this.player.getId() - 1, this.destroyPos, -1));
+                }
+                // Tentacles end
+
                 this.level.destroyBlockProgress(this.player.getId(), pos, -1);
                 this.debugLogging(pos, true, sequence, "aborted destroying");
 
diff --git a/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java b/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
index bcc4c1af0fe153ed87e55706a1a6a5c6e945fec8..d825c918562e740e774e7daa4fd8fee67f271c92 100644
--- a/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/src/main/java/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -3,6 +3,7 @@ package net.minecraft.world.level.block.state;
 
 import com.google.common.collect.ImmutableMap;
 import com.mojang.serialization.MapCodec;
+import it.unimi.dsi.fastutil.Pair; // Tentacles
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
@@ -20,6 +21,7 @@ import net.minecraft.core.Registry;
 import net.minecraft.network.protocol.game.DebugPackets;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer; // Tentacles
 import net.minecraft.tags.FluidTags;
 import net.minecraft.tags.TagKey;
 import net.minecraft.util.Mth;
@@ -885,9 +887,37 @@ public abstract class BlockBehaviour {
         }
 
         public float getDestroyProgress(Player player, BlockGetter world, BlockPos pos) {
-            return this.getBlock().getDestroyProgress(this.asState(), player, world, pos);
+            // Tentacles start - Mining speed API
+            return this.getDestroyProgressPair(player, world, pos).left();
         }
 
+        public Pair<Float, Boolean> getDestroyProgressPair(Player player, BlockGetter world, BlockPos pos) {
+            final var state = this.asState();
+            float f = state.getDestroySpeed(world, pos);
+
+            if (f == -1.0F) {
+                return Pair.of(0.0F, false);
+            }
+
+            if (player instanceof ServerPlayer serverPlayer) {
+                final var speedPair = serverPlayer.getDestroySpeedPair(state);
+                final float toolModifier = (serverPlayer.alwaysCorrectTool || player.hasCorrectToolForDrops(state)) ? 30F : 100F;
+
+                var speed = speedPair.left() / f / toolModifier;
+                // For some reason when there is a modifier enabled, and a tool has > 1 speed,
+                // The actual mining speed is too high, this fixes that. (This took like 3 hours to figure out)
+                if (speedPair.right() && toolModifier == 30 && player.getMainHandItem().getDestroySpeed(state) > 1F) {
+                    speed *= 0.3F;
+                }
+
+                return Pair.of(speed, speedPair.right() || serverPlayer.alwaysCorrectTool);
+            }
+
+            final float toolModifier = player.hasCorrectToolForDrops(state) ? 30F : 100F;
+            return Pair.of(player.getDestroySpeed(state) / f / toolModifier, false);
+        }
+        // Tentacles end
+
         public int getDirectSignal(BlockGetter world, BlockPos pos, Direction direction) {
             return this.getBlock().getDirectSignal(this.asState(), world, pos, direction);
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index bd363cd808f294f47298258ee7b99aeb36150887..fe7b5a3af027bbd4697114903a54921ac568f855 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3094,4 +3094,16 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         this.getHandle().connection.send(packet);
     }
     // Purpur end
+
+    // Tentacles start
+    @Override
+    public boolean isAlwaysCorrectTool() {
+        return getHandle().alwaysCorrectTool;
+    }
+
+    @Override
+    public void setAlwaysCorrectTool(boolean alwaysCorrectTool) {
+        getHandle().alwaysCorrectTool = alwaysCorrectTool;
+    }
+    // Tentacles end
 }
