From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DaRacci <racci@racci.dev>
Date: Fri, 3 Jun 2022 20:30:55 +1000
Subject: [PATCH] Mining speed API


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 7344324517fb10ee6ab3467b40fcb9cd798961cb..f8e458911535040ea58382b693601634d102e72a 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -189,7 +189,10 @@ public abstract class Player extends LivingEntity {
     public int sixRowEnderchestSlotCount = -1; // Purpur
     public boolean flyingFallDamage = false; // Purpur
     public boolean canPortalInstant = false; // Purpur
-
+    // Tentacles start
+    public boolean normalUnderwaterMining = false;
+    public boolean normalOffGroundMining = false;
+    // Tentacles end
     // CraftBukkit start
     public boolean fauxSleeping;
     public int oldLevel = -1;
@@ -875,11 +878,11 @@ public abstract class Player extends LivingEntity {
             f *= f1;
         }
 
-        if (this.isEyeInFluid(FluidTags.WATER) && !EnchantmentHelper.hasAquaAffinity(this)) {
+        if (!normalUnderwaterMining && this.isEyeInFluid(FluidTags.WATER) && !EnchantmentHelper.hasAquaAffinity(this)) { // Tentacles
             f /= 5.0F;
         }
 
-        if (!this.onGround) {
+        if (!normalOffGroundMining && !this.onGround) { // Tentacles
             f /= 5.0F;
         }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 619e73ff2270bbfab93c11dabdae95c601362342..81f224b6017cd6b2bf84d429a6ac854fc670de37 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2958,4 +2958,26 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         return getHandle().flyingFallDamage;
     }
     // Purpur end
+
+    // Tentacles start
+    @Override
+    public boolean isNormalUnderwaterMining() {
+        return getHandle().normalUnderwaterMining;
+    }
+
+    @Override
+    public boolean isNormalOffGroundMining() {
+        return getHandle().normalOffGroundMining;
+    }
+
+    @Override
+    public void setNormalUnderwaterMining(boolean normalUnderwaterMining) {
+        this.getHandle().normalUnderwaterMining = normalUnderwaterMining;
+    }
+
+    @Override
+    public void setNormalOffGroundMining(boolean normalOffGroundMining) {
+        this.getHandle().normalOffGroundMining = normalOffGroundMining;
+    }
+    // Tentacles end
 }
